include/exadg/solvers_and_preconditioners/utilities/petsc_operation.h

```diff
@@ -30,36 +30,29 @@ using namespace dealii;
 
 /*
  *  This function wraps the copy of a PETSc object (sparse matrix,
- *  preconditioner) with a dealii::LinearAlgebra::distributed::Vector.
+ *  preconditioner) with a dealii::LinearAlgebra::distributed::Vector, taking
+ *  pre-allocated PETSc vector objects for the temporary operations
  */
 #ifdef DEAL_II_WITH_PETSC
 template<typename VectorType>
 void
 apply_petsc_operation(VectorType &                                           dst,
                       VectorType const &                                     src,
                       std::function<void(PETScWrappers::VectorBase &,
-                                         PETScWrappers::VectorBase const &)> petsc_operation)
+                                         PETScWrappers::VectorBase const &)> petsc_operation,
+                      Vec &                                                  petsc_vector_dst,
```

Comment:
I wonder if the wrapping idea of this function would become more clear if the PETSc vectors would come as third and fourth arguments, and finally the lambda?

As a general question: Where does the type `Vec` come from? Did we define this in ExaDG?

Reply:
I agree, I moved the lambda to the last argument. I also put the lambda last in the other `apply_petsc_operation` for consistency (even though it is only a communicator there), see fc0d323d.

The type `Vec` is the name of the PETSc struct, https://www.mcs.anl.gov/petsc/petsc-current/docs/manualpages/Vec/Vec.html

Reply:
Ok, I just thought it looks strange that `Vec` is written without any namespace. We could think about `typedef Vec VectorTypePETSc;` with a short comment, and write explicitly `dealii::PETScWrappers::VectorBase` here in order to be more explicit about the types and where they come from (it might get quite complicated for users without dealii background)? It probably does not make sense to wrap the vector into `PETScWrappers::VectorBase` outside of this function (because this is the main content of `apply_petsc_operation`).

Reply:
Very good idea. Done in 0f0e2b69.